airflow/io/path.py

```diff
@@ -173,10 +176,14 @@ def bucket(self) -> str:
     @property
     def key(self) -> str:
         if self._url:
-            return self._url.path
+            return self._url.path.lstrip(self.sep)
```

Comment:
Is this a bug fix? If so, moving it to a separate PR would be better, because it looks like a breaking change in this PR.

Reply:
Somewhat, but there is nothing that relies on it. I need it to be equal to what S3 provides and to deliver equal results. I need it in this PR though, otherwise I need to add an additional 'bugfix'. I dont think that makes sense.

Reply:
Lstrip a path doesn’t sound right to me; I’d instead prefer ensuring a slash is present in the front. (Although tachnically the two probably are equivalent since this is just a key and not actually semantically need to be a valid path…)

Reply:
On the other hand, should we do a rstrip? That’s what Python’s pathlib does, and while a trailing slash is technically significant in a URI, removing it seems to make sense here since only one of the variants can make sense for any particular existing resource.

Reply:
It depends how we want the key to an object to look look like

What will this do:

```
s3://my-bucket/my/key

bucket = my-bucket  # notice no '/'
key = "my/key" 

# or 

key = "/my/key"
```

The S3 hook (https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_modules/airflow/providers/amazon/aws/hooks/s3.html#S3Hook.get_s3_bucket_key) is explicit about making it a relative path (so with the initial '/').

I agree with adding a rstrip or then basically a full `strip`.

Reply:
I thin we should have some flexibility here. There is at least local filesystem where semantic is different depending on presence or lack of the leading slash. Specifically local filesystem makes a difference:

```
file:///tmp/test.json
```

is different tham:

```
file://tmp/test.json
```

The latter is a file in temporary subfolder in current working directory, where the former is "/tmp/test.json" starting from the root. I think stripping behaviour should depend on the filesystem.

Reply:
Oh yes, lstrip removes all slashes while only 1 needs to be removed. I do prefer not to distinguish based of filesystem, as we call our implementation ObjectStoragePath and document it to have object storage semantics.

But indeed here I think it needs to be

```python
    if s.startswith('/'):
        s = s[1:]
```

Reply:
I am not sure - what's the semantic of key vs. path. What's the "semantic" difference between those I wonder?

Reply:
disclaimer: i'm not sure what is right here either.

Given that documentation on object stores like [[1]](https://docs.openstack.org/api-ref/object-store/#create-or-replace-object) [[2]](https://www.educative.io/answers/what-do-keys-mean-in-amazon-s3) [[3]](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html) [[4]](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blobs-list) all use *relative* identifiers to objects/keys and they construct an URI to to an object as follows:

```python
uri = f"{protocol}://{container}/{object}" # note the hardcoded '/'
# or in s3 speak
uri = f"{protocol}://{bucket}/{key}" # again the hardcoded '/'
```

Double slashes don't make sense here because if referencing a local file the bucket would be "" (null/empty):

```python
b = ObjectStoragePath("file:///tmp/data.json")
assert b.bucket == ""
```

Reconstructing the uri would then still work:

```python

uri = f"{b.protocol}://{b.bucket}/{b.path}"  # uri = "file:///tmp/data.json"

```

I think the proper way to reference a key/object is relative and thus stripping a slash. 

`path` is obtained from the urlsplit and we always assume a non-relative url that is passed to ObjectStoragePath. So considering that I think stripping the leading slashes from a `key` is the correct behavior.

Reply:
Furthermore given https://en.wikipedia.org/wiki/File_URI_scheme a format of "file://tmp/file" is actually not valid when omitting the host, ie. it always needs to be an absolute reference.

Reply:
Local storage is inherently unreliable when talking about lineage :) 
It would definitely be the best to have absolute references though.